### ISSUE TITLE
Win10 custom scaling causes text to move fractionally when scrolling/typing

### ISSUE THREAD

**Issue by qgates** [microsoft/vscode#117140]
- VS Code Version: 1.53.2
- OS Version: Win10

This issue appears to be a quantisation / calculation error painting text lines in editor panes, which occurs at custom scaling levels in Win10. At 100% (normal) scaling, the issue does not appear.

Interestingly this issue started to appear at more recent builds, possibly since the Electron update (or possibly just before). Code has been used on the same rig at 110% scaling April-December 2020 without any issue and the problem started to appear recently.

Steps to Reproduce:

1. Set a custom scaling of 110% (in Win10 Display.. scale and layout.. advanced scaling settings)
2. Scroll to different parts of a document and interact with text via mouse clicks, selection or typing
3. Observe that when beginning to type or move the caret with LR cursor keys, the current text line may move up or down by a pixel. Moving UD with cursor keys results in the same movement on the line being moved to.
4. Possibly related, doing a column select reveals inconsistent inter-line gaps, consistent at normal scaling

Does this issue occur when all extensions are disabled?: Yes

**Comment by alexdima** [comment]
@qgates Would the following help: closing all instances of vscode and executing `code --force-device-scale-factor=1`? That should make it that the Windows configured scale is ignored by Electron/Chromium. Then you could potentially increase the fontSize in the editor to a good value. #519 tracks that the font size of other workbench elements could be tweaked.

The problem is that this scaling is done outside of vscode. We position everything at integer coordinates, but then Chromium's scaling kicks in and due to rounding errors causes these gaps. I could not find any clear workaround on our side except that we would own the "zooming" like we can with the fontSize and lineHeight...

**Comment by qgates** [comment]
@alexdima In case you missed it I modified my original post to reflect the fact that the problem occurs **as well** when using `window.zoomLevel` (ie. Chromium's canvas zoom). I see the issue whether using Win10 scaling ***or*** window.zoomLevel, or a combination of both.

That suggests (to me at least) that there is some rounding error with the calculation of paint positions, and that sometimes the error falls on either side of the rounding.

Another way to see the problem is to load up a large, long line of text with wrapping enabled, then search for a common string and hit F3 to repeat the search (minified CSS is good for this test). Often when the highlight moves to the next match on a new line, the line will shift down or up by a pixel. And that's without Win10 scaling, just with window.zoomLevel set to 0.3.

If neither Win10 scaling nor window.zoomLevel are being used, the problem doesn't appear.

If I try your suggestion I would end up needing to use window.zoomLevel anyway to make other elements of vscode big enough (sidebar font, tabs font etc.). I am using a 27" 2K screen, and everything's too small at 1:1 scaling.

I'd also be curious to know why this issue started appearing at around December, where everything appeared to be working correctly beforehand.

**Comment by alexdima** [comment]
@qgates I agree with you, any Chromium scaling might reproduce the problem. I acknowledge the bug and am aware about it since >5 years. The bug exists in Chromium's way of doing scaling. Basically two segments [0->19),[19->38) will no longer be touching once they do the scaling under certain zoom levels.

AFAIK there is nothing for us to do on our side. The only way to get pixel perfect rendering is to disable Chromium's scaling (either induced by the `window.zoomLevel` setting or induced by the Windows Device Scaling setting) or to use 150%, 200%, or some zoom level where the rounding errors are not so bad.

Once Chromium's scaling is disabled, `editor.fontSize` can be used to increase the editor's font size. The rest of the workbench cannot be logically increased until #519 is done, which is why I encourage you to upvote that.

**Comment by qgates** [comment]
@alexdima Thanks for your explanation and time. I took a look at #8606 and though it's not the same issue it would seem as though the what I describe is related, ie. is some kind of rounding / quantising error introduced by scaling whether by Win10 or Chromium.

Out of interest have you managed to replicate or see the exact issue I'm describing? Ie. odd groups of lines moving up or down by a pixel when moving the caret on to them, editing and/or selecting (eg. with F3-search). Reason for asking is that I'm fairly perceptive to such things and I could swear that I only started to notice it around December. It is of course possible that it's been around for years and that I didn't notice before, but that's not typically me 😉

I can live with running vscode and Win10 at standard scaling and use fontSize and lineHeight to make things workable, but my eyesight started degrading a few years ago so I opted to use 110% (slight) upscale in Win10 which makes everything nicer on a 27" 2560x1440 panel, together with 0.3 zoomLevel which made everything about perfect. But in the absence of Win10/Chromium scaling, adjusting other UI element in vscode would get me to the same place so #519 upvoted 🙂

**Comment by alexdima** [comment]
@qgates You could also try the other way around. e.g. 
* close all instances of vscode
* execute ` code --force-device-scale-factor=1` (this will instruct Chromium to ignore the multiplier scale coming in from Win10)
* configure `window.zoomLevel: 2` or 3 and then reduce the fontSize and lineHeight. It could be that one of these zoom factors results in a better zoom factor with smaller (or more rare) rounding errors given how floats are represented.
